src/prompts/headless/manager.ts

```diff
@@ -0,0 +1,318 @@
+import type {
+  __Autogenerated__PromptId,
+  __Autogenerated__PromptMajorVersion,
+  __Autogenerated__PromptMinorVersion,
+  __Autogenerated__PromptModelParams,
+  __Autogenerated__PromptTemplateId,
+  __Autogenerated__PromptTemplateParams,
+} from '../autogenerated';
+import {
+  zHeadlessPromptSchema,
+  type TimeDelta,
+  type PromptTracking,
+  type HeadlessPrompt,
+} from '../../types';
+import {
+  readEnv,
+  AutoblocksEnvVar,
+  convertTimeDeltaToMilliSeconds,
+  HEADLESS_PROMPT_LATEST_VERSION,
+} from '../../util';
+import { renderTemplate } from '../util';
+
+export class AutoblocksHeadlessPromptManager<
+  PromptId extends __Autogenerated__PromptId,
+  MajorVersion extends __Autogenerated__PromptMajorVersion<PromptId>,
+  MinorVersion extends __Autogenerated__PromptMinorVersion<
+    PromptId,
+    MajorVersion
+  >,
+> {
+  private readonly id: PromptId;
+  private readonly majorVersion: string;
+  private readonly minorVersion: string | { version: string; weight: number }[];
+  private readonly minorVersionsToRequest: string[];
+
+  private readonly apiKey: string;
+  private readonly apiBaseUrl: string = 'https://api.autoblocks.ai';
+
+  // Map of minor version -> prompt
+  private prompts: Record<string, HeadlessPrompt> = {};
+
+  private readonly refreshIntervalTimer: NodeJS.Timer | undefined;
+  private readonly refreshTimeout: TimeDelta;
+  private readonly initTimeout: TimeDelta;
+
+  constructor(args: {
+    id: PromptId;
+    version: {
+      major: MajorVersion;
+      minor:
+        | MinorVersion
+        | [
+            { version: MinorVersion; weight: number },
+            ...{ version: MinorVersion; weight: number }[],
+          ];
+    };
+    apiKey?: string;
+    refreshInterval?: TimeDelta;
+    refreshTimeout?: TimeDelta;
+    initTimeout?: TimeDelta;
+  }) {
+    this.id = args.id;
+    this.majorVersion = args.version.major;
+    this.minorVersion = args.version.minor;
+    this.minorVersionsToRequest = makeMinorVersionsToRequest(this.minorVersion);
+
+    const apiKey = args.apiKey || readEnv(AutoblocksEnvVar.AUTOBLOCKS_API_KEY);
+    if (!apiKey) {
+      throw new Error(
+        `You must either pass in the API key via 'apiKey' or set the '${AutoblocksEnvVar.AUTOBLOCKS_API_KEY}' environment variable.`,
+      );
+    }
+    this.apiKey = apiKey;
+
+    if (
+      this.minorVersionsToRequest.some(
+        (version) => version === HEADLESS_PROMPT_LATEST_VERSION,
+      )
+    ) {
+      const refreshInterval = args.refreshInterval || { seconds: 10 };
+      console.log(
```

Comment:
I think a future enhancement that would probably be nice is the ability to pass in a logger